back/engines/commercial/custom_maps/app/controllers/custom_maps/web_api/v1/layers_controller.rb

```diff
@@ -52,21 +52,34 @@ def reorder
 
         private
 
-        def layer_params
+        def create_params
+          layer_params(base_permitted_params << :type)
+        end
+
+        def update_params
+          layer_params(base_permitted_params)
+        end
+
+        def layer_params(permitted_params)
           params.require(:layer)
-            .permit(
-              :default_enabled,
-              :marker_svg_url,
-              geojson_file: %i[filename base64],
-              title_multiloc: CL2_SUPPORTED_LOCALES
-            ).tap do |whitelisted|
+            .permit(permitted_params).tap do |whitelisted|
             whitelisted[:geojson] = params.dig(:layer, :geojson)
             whitelisted.permit!
           end
         end
 
+        def base_permitted_params
+          [
+            :layer_url,
+            :default_enabled,
+            :marker_svg_url,
+            { geojson_file: %i[filename base64] },
+            { title_multiloc: CL2_SUPPORTED_LOCALES }
+          ]
+        end
+
         def serialized_layer
-          CustomMaps::WebApi::V1::LayerSerializer.new(@layer).serializable_hash.to_json
+          "CustomMaps::WebApi::V1::#{@layer.type.demodulize}Serializer".constantize.new(@layer).serializable_hash.to_json
```

Comment:
This is considered a little bit dangerous. It's very hypothetical, but suppose there would exist a `DeleteAllDataOnInstantiationSerializer` then this class could get instantiated by passing a corresponding value for the type in the API call. In practice there is probably no danger at all. But it could be worth considering to instead write something like:
```
serializer_class = case @layer.type
  GeoJsonSerializer when 'GeoJSON'
  EsriSerializer when 'ESRI'
  end
```

Reply:
Good idea. I was aware of the issue of malicious class calling via the API, but thought the whitelisting of the `type` in the `Layer` model was sufficient to prevent it.

I'll add your conditional idea into the method, however, for extra security 👍